### ISSUE TITLE
Simple Android Implementation Question

### ISSUE THREAD

**Issue by thertzelle** [tangrams/tangram-es#2253]
I was wondering if anyone that had some Android experience and Tangram-ES experience could take a look at this, or give me a hint? I've been trying to get this to work for 48 hours straight and not seeing what I'm doing wrong here.

Below is my code and below that is the stacktrace that happens when I run it. If anyone has any ideas, I would greatly appreciate it.

```class TangramMapView(private val context: ThemedReactContext, private val appContext: ReactApplicationContext): FrameLayout(appContext), MapView.MapReadyCallback, LifecycleEventListener {

  private var _map: MapController? = null
  private var _view: MapView? = null

  fun onAfterUpdateTransaction() {

    if (_view != null) {
      return;
    }

    _view = MapView(this.getContext())

    val mapViewLayoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    this.addView(_view, 0, mapViewLayoutParams)

    _view?.onCreate(null)
    _view?.getMapAsync(this)

    appContext.addLifecycleEventListener(this)
  }

  override fun onMapReady(mapController: MapController?) {
    _map = mapController

    _map?.loadSceneFileAsync("pref-style.yaml")
  }

  override fun onHostResume() {
    _view?.onResume()
  }

  override fun onHostPause() {
    _view?.onPause()
  }

  override fun onHostDestroy() {
    _view?.onDestroy()
  }
}
```

```
D/Tangram: Loading native library took 58ms
D/Tangram: Loading native library took 0ms
D/Tangram: MapController creation took 3ms
D/Tangram: MapController init took 7ms
D/Tangram: WARNING tileManager.cpp:204: add source _mine
E/Tangram: Error in font.xml parsing: END_TAG expected (position:START_TAG (empty) <axis tag='wght' stylevalue='400'>@153:49 in java.io.InputStreamReader@9bd1ba6) 
E/Tangram: Error in font.xml parsing: END_TAG expected (position:START_TAG (empty) <axis tag='wdth' stylevalue='100.0'>@292:51 in java.io.InputStreamReader@9bd1ba6) 
E/Tangram: Error in font.xml parsing: END_TAG expected (position:START_TAG (empty) <axis tag='wght' stylevalue='400'>@369:49 in java.io.InputStreamReader@9bd1ba6) 
D/Tangram: FontConfig init took 49ms
A/libc: Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 9577 (Thread-6), pid 9513 (ctnativetangram)
```

**Comment by matteblair** [comment]
Hi @thertzelle! The code you've posted here doesn't look like it has any obvious problems. In the logcat output from your app, do you see any native crash dumps like the ones listed here: https://source.android.com/devices/tech/debug ? If you can share a crash dump or tombstone that would make it much easier to diagnose the issue.

**Comment by matteblair** [comment]
Oh and also, can you share the device and OS version that you're seeing this issue on? Thanks.

**Comment by thertzelle** [comment]
Thank you for your reply. I'm running on MacOS and Emulator: Pixel_3a_API_30_x86

Here is the crash dump:

```
    --------- beginning of crash
2021-03-21 22:07:33.489 15619-15708/com.example.reactnativetangram A/libc: Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 15708 (Thread-6), pid 15619 (ctnativetangram)
2021-03-21 22:07:33.532 15715-15715/? I/crash_dump32: obtaining output fd from tombstoned, type: kDebuggerdTombstone
2021-03-21 22:07:33.533 284-284/? I/tombstoned: received crash request for pid 15708
2021-03-21 22:07:33.534 15715-15715/? I/crash_dump32: performing dump of process 15619 (target tid = 15708)
2021-03-21 22:07:33.541 15715-15715/? A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
2021-03-21 22:07:33.541 15715-15715/? A/DEBUG: Build fingerprint: 'google/sdk_gphone_x86_arm/generic_x86_arm:11/RSR1.201013.001/6903271:userdebug/dev-keys'
2021-03-21 22:07:33.541 15715-15715/? A/DEBUG: Revision: '0'
2021-03-21 22:07:33.541 15715-15715/? A/DEBUG: ABI: 'x86'
2021-03-21 22:07:33.542 15715-15715/? A/DEBUG: Timestamp: 2021-03-21 22:07:33-0500
2021-03-21 22:07:33.542 15715-15715/? A/DEBUG: pid: 15619, tid: 15708, name: Thread-6  >>> com.example.reactnativetangram <<<
2021-03-21 22:07:33.542 15715-15715/? A/DEBUG: uid: 10153
2021-03-21 22:07:33.542 15715-15715/? A/DEBUG: signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
2021-03-21 22:07:33.542 15715-15715/? A/DEBUG:     eax 00000000  ebx 00003d03  ecx 00003d5c  edx 00000006
2021-03-21 22:07:33.542 15715-15715/? A/DEBUG:     edi edb9d81e  esi bdfe7770
2021-03-21 22:07:33.542 15715-15715/? A/DEBUG:     ebp f23d1b90  esp bdfe7718  eip f23d1b99
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG: backtrace:
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #00 pc 00000b99  [vdso] (__kernel_vsyscall+9)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #01 pc 0005ad68  /apex/com.android.runtime/lib/bionic/libc.so (syscall+40) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #02 pc 00076511  /apex/com.android.runtime/lib/bionic/libc.so (abort+209) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #03 pc 000b8d4c  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libc++_shared.so (BuildId: 8771d1ad0fbd042529aa00a1937ddbc7b237a107)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #04 pc 000b29d0  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libc++_shared.so (__gxx_personality_v0+416) (BuildId: 8771d1ad0fbd042529aa00a1937ddbc7b237a107)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #05 pc 0003d049  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libjscexecutor.so (BuildId: 6d04be91b5cf852c8e29737aa4225fe6ee02a9b0)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #06 pc 0003d33b  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libjscexecutor.so (_Unwind_RaiseException+263) (BuildId: 6d04be91b5cf852c8e29737aa4225fe6ee02a9b0)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #07 pc 000b6d6c  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libfbjni.so (__cxa_throw+108) (BuildId: 3546c1351f83366cb88b3802c4c63ee907e2db98)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #08 pc 0005c7a4  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libc++_shared.so (std::__ndk1::locale::use_facet(std::__ndk1::locale::id&) const+212) (BuildId: 8771d1ad0fbd042529aa00a1937ddbc7b237a107)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #09 pc 000d40b1  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libtangram.so (BuildId: 6029bb1d2461bd379ef7d9feb45290473685c5ee)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #10 pc 001887c3  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libtangram.so (BuildId: 6029bb1d2461bd379ef7d9feb45290473685c5ee)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #11 pc 000a779d  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libtangram.so (BuildId: 6029bb1d2461bd379ef7d9feb45290473685c5ee)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #12 pc 0003f834  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libtangram.so (BuildId: 6029bb1d2461bd379ef7d9feb45290473685c5ee)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #13 pc 000304bb  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libtangram.so (BuildId: 6029bb1d2461bd379ef7d9feb45290473685c5ee)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #14 pc 000306fd  /data/app/~~n5hkrH_D2pLOLJ8v0_txqg==/com.example.reactnativetangram-5EAqjoO3_VP-jZreV89Pug==/lib/x86/libtangram.so (BuildId: 6029bb1d2461bd379ef7d9feb45290473685c5ee)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #15 pc 000e6974  /apex/com.android.runtime/lib/bionic/libc.so (__pthread_start(void*)+100) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)
2021-03-21 22:07:33.544 15715-15715/? A/DEBUG:       #16 pc 00078567  /apex/com.android.runtime/lib/bionic/libc.so (__start_thread+71) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)
```

**Comment by matteblair** [comment]
Thanks! It looks like you're running Tangram ES in a React Native app - is that right? The stack trace includes some SOs that I'm unfamiliar with. 

Here's a build of the most recent release (0.16.0) that includes debug symbols. Can you replace your tangram AAR with this one and send any resulting crash logs?
[tangram-debug-4fd1806.aar.zip](https://github.com/tangrams/tangram-es/files/6179295/tangram-debug-4fd1806.aar.zip)

**Comment by thertzelle** [comment]
I actually hadn't seen that stack trace before, but seeing the line about `libc++_shared.so` was sort of a clue. Previously I was getting an error: `More than one file was found with OS independent path 'lib/x86/libc++_shared.so'`. Which people on the internet were saying to do this:
```
    packagingOptions {
      pickFirst 'lib/*/libc++_shared.so'
    }
```
So I did that days ago and didn't think much about it since then. Well after googling a bit I ran into [this](https://developer.android.com/ndk/guides/cpp-support#one_stl_per_app) and [this thing](https://www.reddit.com/r/reactnative/comments/l45uib/conflicting_native_shared_object_libc_sharedso/)!

So it seems React Native uses `libc++_shared.so` and it seems Tangram is using the system one, which if you read the first link from above it says something to the effect of, don't do this.

But per the second link's suggestion, I added `SoLoader.loadLibrary("tangram");`

I don't know enough about `SoLoader.loadLibrary` to know what it is doing to prevent some sort of conflict with `libc++_shared.so`, also note that I still needed to include `packagingOptions` from above.
